### PR TITLE
feature/map page

### PR DESCRIPTION
**Pull request by neozhixuan**
- Instantiated the map page
- Used dynamic imports to integrate Leaflet.js components to Next.js
- Utilised react context to load selected items + all locations of facilities
- Utilised existing functions to find the nearest facilities and display based on distance from user location
- Displayed map markers and used cluster function to reduce clutter on screen
- Map markers are differentiated by the image in the icon
- Working filter to choose the items to display as well as customise the distance from user location to display facility markers

- Implementation of a footer that can be pulled up, but archived the function - however, similar code can be used for upcoming implementations of mockup design

### CODE REVIEW

File: client/app-context/UserSelectionContext/types.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ export type AddressOption = {
 	label: string;
 	coordinates: Coordinates;
 };
-export type TResult = { facilities: { id: number; distance: number }[] };
+export type TResult = { facilities: { id: number; distance: number; latlng: Array<number> }[] };
```
Comment (ryhuz) [inline client/app-context/UserSelectionContext/types.ts:15]: i'm imagining that `latlng` will be `[lat,lng]`, and if so, should be typed as such

File: client/components/map/FilterPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+									value={item.value}
+									name={item.method}
+								>
+									{item.value}
+								</Checkbox>
+							))}
+						</Stack>
+					</CheckboxGroup>
+				</VStack>
+				{/* <VStack align="flex-start" w="100%">
```
Comment (ryhuz) [inline client/components/map/FilterPanel.tsx:114]: remove this commented code

File: client/components/map/LeafletMap.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+		center: LatLngExpression;
+		children: JSX.Element | JSX.Element[];
+		zoom: number;
+	} & MapOptions
+> = ({ ...options }) => {
+	const { setMap } = useMapContext();
+	const [isClient, setIsClient] = useState(false);
+
+	useEffect(() => {
+		setIsClient(true);
```
Comment (ryhuz) [inline client/components/map/LeafletMap.tsx:20]: why do we need a state that immediately is set to true?

File: client/components/map/Marker/MarkerIconWrapper.tsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+export interface CustomMarkerProps {
+	position: LatLngExpression;
+	icon?: React.ComponentType<IconProps>; // Update the type here
+	color: string;
+	label?: string;
+	category: string;
+}
+
+const MarkerIconWrapper = ({ icon, color, label, category }: Partial<CustomMarkerProps>) => {
+	const IconFC = useMemo(() => icon ?? null, [icon]);
```
Comment (ryhuz) [inline client/components/map/Marker/MarkerIconWrapper.tsx:18]: don't think this is necessary? we can just use icon direct?

if the concern is that icon is not capitalised, maybe we can use and alias

`{icon: IconFC, ...` in the props on line 17

File: client/components/map/Marker/index.tsx [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	position: CustomMarkerProps["position"];
+	icon: CustomMarkerProps["icon"];
+	color: CustomMarkerProps["color"];
+	handleOnClick: () => void;
+	category?: string;
+}> = ({ position, icon, color, handleOnClick, category }) => {
+	const { map } = useMapContext();
+	const { divIcon } = useLeafletDivIcon();
+
+	// const handleMarkerClick = () => {
```
Comment (ryhuz) [inline client/components/map/Marker/index.tsx:17]: remove if not needed

File: client/components/map/NearbyFacilitiesPanel.tsx [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+						leftIcon={<ChevronLeftIcon />}
+						onClick={() => setPage(Pages.HOME)}
+					>
+						Restart!
+					</Button>
+					<Location showText={false} handleBlur={handleChangedLocation} />
+					<FilterButton onClick={showFilter} />
+				</Flex>
+				<Flex flexDir={"column"} maxW={"100vw"} p={5}>
+					{nearbyLocations.results &&
```
Comment (ryhuz) [inline client/components/map/NearbyFacilitiesPanel.tsx:98]: this whole block looks very deeply nested. is it possible to break it up into components?

File: client/components/map/StickyFooter.tsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+						position="absolute"
+						top="-10px"
+						left="50%"
+						transform="translateX(-50%)"
+						w="30px"
+						h="2px"
+						bg="black"
+					/>
+					<Box>
+						<p>Your Footer Text</p>
```
Comment (ryhuz) [inline client/components/map/StickyFooter.tsx:34]: is there actual footer text that supposed to go here?

File: client/components/map/utils.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+import { useState } from "react";
+export const useForceUpdate = () => {
```
Comment (ryhuz) [inline client/components/map/utils.ts:2]: what's this for?

File: client/leaflet.d.ts [hunk 9]
```diff
@@ -0,0 +1,8 @@
+// leaflet.d.ts
+declare global {
+	interface Window {
+		L: any; // Adjust this type based on Leaflet's actual types if available
```
Comment (ryhuz) [inline client/leaflet.d.ts:4]: maybe can try console logging what it is and typing it as closely as possible

File: client/public/Marker.css [hunk 10]
```diff
@@ -0,0 +1,30 @@
+.leaflet-marker-icon {
+	border: 0cm;
+	background: rgba(0, 0, 0, 0);
+}
+
+.marker-text {
+	font-size: 9px !important;
```
Comment (ryhuz) [inline client/public/Marker.css:7]: is this `!important` important? if possible, can we do away with all of the importants?

**Review by ryhuz — CHANGES_REQUESTED** [review]

File: client/components/map/LeafletMap.tsx (hunk 3 above)
Comment (neozhixuan) [reply client/components/map/LeafletMap.tsx:20]: I think this is to ensure that the map component only loads when the component is mounted in the client side

File: client/components/map/utils.ts (hunk 8 above)
Comment (neozhixuan) [reply client/components/map/utils.ts:2]: It is used to update the state after changing the address, as the map was not changing when I entered a new address. This function exists in the MapPage itself so I will be removing it, but not sure if there is a better implementation for this

File: client/public/Marker.css (hunk 10 above)
Comment (neozhixuan) [reply client/public/Marker.css:7]: Some of them are important as it gets overriden by existing Chakra styles, but I removed the unnecessary ones

File: client/components/map/NearbyFacilitiesPanel.tsx (hunk 6 above)
Comment (neozhixuan) [reply client/components/map/NearbyFacilitiesPanel.tsx:98]: This component is being archived (it was the component to pull up and show nearby facilities, but design changed)

File: client/components/map/StickyFooter.tsx (hunk 7 above)
Comment (neozhixuan) [reply client/components/map/StickyFooter.tsx:34]: Removing this component as it is unused too

File: client/app-context/UserSelectionContext/types.ts (hunk 1 above)
Comment (neozhixuan) [reply client/app-context/UserSelectionContext/types.ts:15]: Will be using LatLngExpression from leaflet

**Review by ryhuz — APPROVED** [review]

**Pull request merged**